src/erlang_ls_indexer.erl

```diff
@@ -0,0 +1,210 @@
+%% TODO: Show progress messages to client
+-module(erlang_ls_indexer).
+
+%% TODO: Move all code in one place
+%% TODO: Rename tables
+-callback index(erlang_ls_document:document()) -> ok.
+
+%% TODO: Solve API mix (gen_server and not)
+%% API
+-export([ find_and_index_file/1
+        , find_and_index_file/2
+        , index/1
+        , index_dir/1
+        , start_link/0
+        , index_app/0
+        , index_deps/0
+        , index_otp/0
+        ]).
+
+%% gen_server callbacks
+-export([ init/1
+        , handle_call/3
+        , handle_cast/2
+        , terminate/2
+        ]).
+
+%%==============================================================================
+%% Includes
+%%==============================================================================
+-include("erlang_ls.hrl").
+
+%%==============================================================================
+%% Types
+%%==============================================================================
+-type state() :: #{}.
+
+%%==============================================================================
+%% Macros
+%%==============================================================================
+-define(SERVER, ?MODULE).
+
+%%==============================================================================
+%% Exported functions
+%%==============================================================================
+-spec find_and_index_file(string()) ->
+   {ok, uri()} | {error, any()}.
+find_and_index_file(FileName) ->
+  find_and_index_file(FileName, async).
+
+-spec find_and_index_file(string(), async | sync) ->
+   {ok, uri()} | {error, any()}.
+find_and_index_file(FileName, SyncAsync) ->
```

Comment:
The `SyncAsync` argument is for being able to request one or the other as a client of the indexer. This is useful when we need "find and index" and then rely on the fact that the file has already been indexed.